js/app/modules/filter/currentSet.js

```diff
@@ -51,25 +51,32 @@ var CurrentSet = new Module.Class({
         return belongs_to_set && !belongs_to_subset;
     },
 
-    // Filter override
-    // This Filter can modify the Xapian query, but it can't express its whole
-    // filtering logic in the Xapian query, so we still need to request more
-    // models in case some returned models must be dropped.
-    can_modify_xapian_query: function () {
-        return false;
-    },
-
     // Filter implementation
     modify_xapian_query_impl: function (query) {
+        let tags = this._current_set.child_tags;
+
+        if (this.invert) {
+            if (query.excluded_tags.indexOf(tags[0]) !== -1)
+                return query;
+        } else {
+            if (query.tags_match_any.indexOf(tags[0]) !== -1)
+                return query;
+        }
+
+        let subtags = [];
+        SetMap.get_children_sets(this._current_set).forEach(set => {
+            subtags = subtags.concat(set.child_tags);
+        });
+
         if (this.invert) {
             return Eknc.QueryObject.new_from_object(query, {
-                excluded_tags: Utils.union(query.excluded_tags,
-                    this._current_set.child_tags),
+                tags_match_any: subtags,
```

Comment:
I think this still has to take any existing `excluded_tags` and `tags_match_any` on the query object into account.

Reply:
Not sure what you mean, can you clarify?

Reply:
The current code in this PR overwrites any `excluded_tags` and `tags_match_any` that were already set on the passed-in query object. So it makes this filter incompatible to chain with other filters that modify those properties. That's the reason why `Utils.union()` was originally in this code.

Reply:
ouch, yeah, my very bad.